### PR TITLE
Introduce an environment variable to change c10 log level

### PR DESCRIPTION
**Pull request by cbalioglu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71746

This PR contains the following improvements:

- It exposes a new environment variable `TORCH_MIN_LOG_LEVEL` that enables users to set the log level of c10 logging facility (supports both GLOG and c10 loggers). Valid values are `INFO`, `WARNING`, `ERROR`, and `FATAL` or their numerical equivalents `0`, `1`, `2`, and `3`.
- It implements an `initLogging()` function and calls it as part of `torch._C` module import to ensure that the underlying logging facility is correctly initialized in Python.

With these changes a user can dynamically set the log level of c10 as in the following example:

```
$ TORCH_CPP_LOG_LEVEL=INFO python my_torch_script.py
```

Differential Revision: [D33756252](https://our.internmc.facebook.com/intern/diff/D33756252/)

### CODE REVIEW

File: c10/util/Logging.cpp
Comment (rohan-varma) [inline c10/util/Logging.cpp:223]: nit: maybe `initTorchLogging` to distinguish it from caffe2? Although maybe not because `InitCaffeLogging` may eventually be deprecated.

File: c10/util/Logging.cpp
Comment (rohan-varma) [inline c10/util/Logging.cpp:277]: Nit: Add a comment explaining how logging will be done when GLOG is not available? i.e. I'm guessing the logging APIs will be from `#include <c10/util/logging_is_not_google_glog.h>` which is included above, but might help future developers understand

**Review by rohan-varma — COMMENTED** [review]
LGTM! Will let the core team take a look and stamp it. Being able to update the log level from python will be quite helpful for OSS users to maximize the benefit of debugability work we've shipped over the last year. And separating `c10::initLogging` from `initCaffeLogging` will help deprecate `initCaffeLogging` in the longer term.

File: c10/util/Logging.h
Comment (malfet) [inline c10/util/Logging.h:320]: Any reason why this must be inline? I.e its probably called one or two times during the program lifetime

File: torch/csrc/Module.cpp
Comment (malfet) [inline torch/csrc/Module.cpp:918]: I don't think we ever expose `c10` as name on Python level, why do it here?
```suggestion
    "set_log_level",
```

File: torch/csrc/Module.cpp
Comment (malfet) [inline torch/csrc/Module.cpp:927]: Same as above
```suggestion
    "get_log_level",
```

File: torch/csrc/Module.cpp
Comment (malfet) [inline torch/csrc/Module.cpp:932]: -EPARSE (not sure what is "threshold for the logger", perhaps we should make it a bit more verbose)
```suggestion
Gets the current log level  for the low level Pytorch logging facility.
```

File: torch/csrc/Module.cpp
Comment (malfet) [inline torch/csrc/Module.cpp:935]: Please consider moving unconditional initialization call towards the start of `initModule` function

**Review by malfet — CHANGES_REQUESTED** [review]
Adding @mruberry to list of reviewers.
Requesting changes as any public API requires a test (which would have immediately revealed, that proposed functions are added not to `torch.` but `torch._C.` module.
Also, I don't think `[get|set]_log_level` belongs at top level pytorch module, but rather to `torch.utils.`

Common convention, in order to make code `mypy`/`pyre` parseable, is to add native functions to `torch._C._func_name`, add function prototype to [torch/_C/__init__.pyi.in](https://github.com/pytorch/pytorch/blob/master/torch/_C/__init__.pyi.in) and then call it from the appropriate module

To be frank, I have slight reservations about this PR, as it seems the only goal of adding those new functions is to expose INFO level logs from c10d, which should be doable by adding `c10::ShowLogInfoToStderr()` call to the c10d logging facility initialization. (as its cumbersome to users to expect them to change log level by setting environment variable and then doing an API call)

Also, is it just me or proposed `torch.set_c10_log_level` feels a bit non-pythonic by accepting just one data type as input.

**Comment by mruberry** [comment]
This is interesting. @cbalioglu would you elaborate more on what you're trying to do here and what scenarios you think are important? Also, for some reference points, do frameworks like JAX, NumPy, and/or TensorFlow expose log levels? If so, how do they do it?

**Comment by cbalioglu** [comment]
> This is interesting. @cbalioglu would you elaborate more on what you're trying to do here and what scenarios you think are important?

Hey @mruberry, right now we are significantly revamping the implementation of several components in PyTorch Distributed, such as our socket class and Store types. As part of this work we are now also logging a lot more information in various verbosity levels which is essential to troubleshoot problems in distributed environments. And to be clear we are using just the regular `LOG(INFO/WARNING/ERROR)` macros of c10, so we do not have a separate logging facility in c10d.

Now the *user-visible* problem we are facing is this: There is no way to change the log level of c10 logging facility from Python in our official PyTorch packages. It defaults to WARNING (held by the `FLAGS_caffe2_log_level` variable), meaning it will always output WARNING and ERROR logs to stderr, but if the user wants to modify this behavior (e.g. output INFO as well, or show only ERROR) they don't have a way to achieve this. We have several customers (e.g. ByteDance) particularly interested in INFO level logs for monitoring and troubleshooting their training jobs who are blocked on this limitation.

This was the *user-visible* problem. In fact technically the problem is slightly more complicated. I assume mostly for historical reasons, there are two versions of c10 logging facility. One is a pretty lightweight custom logger implementation fully contained in c10 itself, and the other one is a slim wrapper on top of GLOG. You can switch between the two implementations by specifying the `USE_GLOG` compile-time flag. If `ON` it will use the version that wraps GLOG, and if `OFF` it will use the custom logger. And by default `USE_GLOG` is `OFF`, meaning our official packages use c10's own lightweight logger.

Having said that regardless of which version you pick, we **do not** offer a way for the user to change the log level from Python. *However* if the GLOG wrapper is used, there is a workaround for this limitation. You can simply set the `GLOG_minloglevel` environment variable to the desired log level and it will "work"; although I don't think anyone considered this as an "official" solution. Unfortunately there is no similar mechanism for c10's own logger. It solely relies on the value of the `FLAGS_caffe2_log_level` variable to determine the log level which we don't expose to Python. So if you are using c10's own logger (which is the default one) you are out of luck.

I believe caffe2 authors were aware of this discrepancy, and came up with the `InitCaffeLogging()` function that consolidated the runtime initialization of both versions. `InitCaffeLogging()` expects a `caffe2_log_level` program argument and *overrides* GLOG's log level with it if GLOG wrapper is used; and uses it as it is if c10's own logger is used. Ultimately `FLAGS_caffe2_log_level` becomes the single source of truth when determining the log level.

This PR ultimately tries to achieve the same solution as `InitCaffeLogging()` for PyTorch. We initialize the logging facility during the import of the `torch._C` module via the `initLogging()` function. However since PyTorch is a library, instead of relying on a program argument, we expect the library user to explicitly set the log level using `torch.set_c10_log_level()` in Python; and if not called, we default to WARNING as before. We also abstract away the remnants of caffe2. Although internally we still rely on the `FLAGS_caffe2_log_level` variable to hold the log level, we do not expose it to the user.

>  Also, for some reference points, do frameworks like JAX, NumPy, and/or TensorFlow expose log levels? If so, how do they do it?

caffe2: [`caffe2_log_level`](https://github.com/pytorch/pytorch/blob/master/c10/util/Logging.cpp#L201)
TensorFlow v1: [`TF_CPP_MIN_LOG_LEVEL`](https://github.com/tensorflow/tensorflow/blob/0b6b491d21d6a4eb5fbab1cca565bc1e94ca9543/tensorflow/core/platform/default/logging.cc#L305) and [`tf.logging.set_verbosity()`](https://www.tensorflow.org/api_docs/python/tf/compat/v1/logging/set_verbosity)
TensorFlow v2: [`TF_CPP_MIN_LOG_LEVEL`](https://github.com/tensorflow/tensorflow/blob/0b6b491d21d6a4eb5fbab1cca565bc1e94ca9543/tensorflow/core/platform/default/logging.cc#L305) and [`tf.get_logger()`](https://www.tensorflow.org/api_docs/python/tf/get_logger)
XGBoost: [`xgb.set.config(verbosity)`](https://xgboost.readthedocs.io/en/latest/parameter.html)
scikit-learn: `verbose` parameter per algorithm

**Comment by cbalioglu** [comment]
> Requesting changes as any public API requires a test (which would have immediately revealed, that proposed functions are added not to torch. but torch._C. module.
Also, I don't think [get|set]_log_level belongs at top level pytorch module, but rather to torch.utils.

> Common convention, in order to make code mypy/pyre parseable, is to add native functions to torch._C._func_name, add function prototype to torch/_C/init.pyi.in and then call it from the appropriate module

Makes sense. I can revise the PR with those changes.

> To be frank, I have slight reservations about this PR, as it seems the only goal of adding those new functions is to expose INFO level logs from c10d, which should be doable by adding c10::ShowLogInfoToStderr() call to the c10d logging facility initialization. (as its cumbersome to users to expect them to change log level by setting environment variable and then doing an API call)

Do you mind opening up a bit? Which environment variable are you referring to?

And just to make sure we are on the same page, I assume you meant *c10* instead of *c10d*. We don't have any logging facility in c10d. We simply use the existing `LOG(INFO/WARNING/ERROR)` macros.

File: torch/csrc/Module.cpp
Comment (cbalioglu) [reply torch/csrc/Module.cpp:918]: My intention was to avoid confusion between Python level logging versus C++ level logging. This function only changes the log level of the C++ logger. I don't have a strong opinion on this, but I still think that being a bit more specific in its name might help users. As an end-user when I call a function named `torch.set_log_level()` I would expect it to have a global effect. However since many of our Python APIs use the `logging` module as well which don't respect this setting, it might get confusing.

File: c10/util/Logging.h
Comment (cbalioglu) [reply c10/util/Logging.h:320]: No particular reason, simply because it was short and closely related to `getLogLevel` which is inline. I can move it to the implementation file in the next revision.

File: c10/util/Logging.cpp
Comment (cbalioglu) [reply c10/util/Logging.cpp:223]: Yes, that's why I kept it `initLogging()`. If in the future `InitCaffeLogging` eventually gets deleted, having `Torch` in its name might sound redundant.

**Comment by mruberry** [comment]
> > This is interesting. @cbalioglu would you elaborate more on what you're trying to do here and what scenarios you think are important?
> 
> Hey @mruberry, right now we are significantly revamping the implementation of several components in PyTorch Distributed, such as our socket class and Store types. As part of this work we are now also logging a lot more information in various verbosity levels which is essential to troubleshoot problems in distributed environments. And to be clear we are using just the regular `LOG(INFO/WARNING/ERROR)` macros of c10, so we do not have a separate logging facility in c10d.

Cool.

> Now the _user-visible_ problem we are facing is this: There is no way to change the log level of c10 logging facility from Python in our official PyTorch packages. It defaults to WARNING (held by the `FLAGS_caffe2_log_level` variable), meaning it will always output WARNING and ERROR logs to stderr, but if the user wants to modify this behavior (e.g. output INFO as well, or show only ERROR) they don't have a way to achieve this. We have several customers (e.g. ByteDance) particularly interested in INFO level logs for monitoring and troubleshooting their training jobs who are blocked on this limitation.

Sounds like a real problem.

> This was the _user-visible_ problem. In fact technically the problem is slightly more complicated. I assume mostly for historical reasons, there are two versions of c10 logging facility. One is a pretty lightweight custom logger implementation fully contained in c10 itself, and the other one is a slim wrapper on top of GLOG. You can switch between the two implementations by specifying the `USE_GLOG` compile-time flag. If `ON` it will use the version that wraps GLOG, and if `OFF` it will use the custom logger. And by default `USE_GLOG` is `OFF`, meaning our official packages use c10's own lightweight logger.

The multiple loggers is an unfortunate complication. Do you think one is preferable to the other?

> Having said that regardless of which version you pick, we **do not** offer a way for the user to change the log level from Python. _However_ if the GLOG wrapper is used, there is a workaround for this limitation. You can simply set the `GLOG_minloglevel` environment variable to the desired log level and it will "work"; although I don't think anyone considered this as an "official" solution. Unfortunately there is no similar mechanism for c10's own logger. It solely relies on the value of the `FLAGS_caffe2_log_level` variable to determine the log level which we don't expose to Python. So if you are using c10's own logger (which is the default one) you are out of luck.

Should we thinking about standardizing logging with GLOG? How disruptive would that be, do you think, @cbalioglu, @malfet? 

Why is setting the environment variable not considered an official solution? Wouldn't that be what GLOG itself recommends doing? Are there scenarios where part of a Python program wants to run in one log level and part in another that would suggest flipping the switch in Python itself is preferable?


> I believe caffe2 authors were aware of this discrepancy, and came up with the `InitCaffeLogging()` function that consolidated the runtime initialization of both versions. `InitCaffeLogging()` expects a `caffe2_log_level` program argument and _overrides_ GLOG's log level with it if GLOG wrapper is used; and uses it as it is if c10's own logger is used. Ultimately `FLAGS_caffe2_log_level` becomes the single source of truth when determining the log level.

That seems kinda convoluted, no?


> This PR ultimately tries to achieve the same solution as `InitCaffeLogging()` for PyTorch. We initialize the logging facility during the import of the `torch._C` module via the `initLogging()` function. However since PyTorch is a library, instead of relying on a program argument, we expect the library user to explicitly set the log level using `torch.set_c10_log_level()` in Python; and if not called, we default to WARNING as before. We also abstract away the remnants of caffe2. Although internally we still rely on the `FLAGS_caffe2_log_level` variable to hold the log level, we do not expose it to the user.

I get that. I guess my questions are more like: should we use the GLOG environment variable if that's a standard? Should we try and standardize on one logger?

> > Also, for some reference points, do frameworks like JAX, NumPy, and/or TensorFlow expose log levels? If so, how do they do it?

> caffe2: [`caffe2_log_level`](https://github.com/pytorch/pytorch/blob/master/c10/util/Logging.cpp#L201) TensorFlow v1: [`TF_CPP_MIN_LOG_LEVEL`](https://github.com/tensorflow/tensorflow/blob/0b6b491d21d6a4eb5fbab1cca565bc1e94ca9543/tensorflow/core/platform/default/logging.cc#L305) and [`tf.logging.set_verbosity()`](https://www.tensorflow.org/api_docs/python/tf/compat/v1/logging/set_verbosity) TensorFlow v2: [`TF_CPP_MIN_LOG_LEVEL`](https://github.com/tensorflow/tensorflow/blob/0b6b491d21d6a4eb5fbab1cca565bc1e94ca9543/tensorflow/core/platform/default/logging.cc#L305) and [`tf.get_logger()`](https://www.tensorflow.org/api_docs/python/tf/get_logger) XGBoost: [`xgb.set.config(verbosity)`](https://xgboost.readthedocs.io/en/latest/parameter.html) scikit-learn: `verbose` parameter per algorithm

Looks like JAX relies on the same environment variables as TensorFlow does: https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html?highlight=logging#low-level-details-and-debugging. 

Thanks for the thorough and interesting writeup, @cbalioglu. Given all that, maybe unifying behind one logger is a little too much today, but could we create a PYTORCH_MIN_LOG_LEVEL environment variable that controlled logging behavior whether GLOG or the C10 logger was used and promote it as our official solution? Would that be future-proof if we decided to unify on one logger or the other in the future? Would it meet customer requirements? And would it be simpler than exposing a Python interface?

From my understanding of this PR (and please help me out with this) I believe that's your proposal, except with an environment variable whose name reflects the project instead of the underlying library (which is just an implementation detail that could be refactored away). 

In the future it might be interesting to invest more in logger UX but we don't have to do that now.

**Comment by cbalioglu** [comment]
> The multiple loggers is an unfortunate complication. Do you think one is preferable to the other? Should we start thinking about standardizing logging with GLOG? How disruptive would that be, do you think, @cbalioglu, @malfet?

Well, GLOG is broadly used, well-tested, and feature rich. So from a pure technical standpoint I would say GLOG takes the lead. However my understanding is that due to its dependency on several other Google libraries we have had users that had trouble using it along with PyTorch. For instance see [this comment](https://github.com/pytorch/pytorch/issues/71660#issuecomment-1020777358). So migrating to a GLOG-only solution might be a risky approach from a UX perspective.

> Why is setting the environment variable not considered an official solution? Wouldn't that be what GLOG itself recommends doing?

I would say if our only logging system was GLOG, we could have considered `GLOG_minloglevel` an official solution. However since you cannot know whether a PyTorch package is built with GLOG or not, you cannot rely on `GLOG_minloglevel`. It might work or not work depending on the compile-time value of `USE_GLOG`.

> That seems kinda convoluted, no?

Yes, it certainly is. This is a complexity caused by supporting two completely different logging systems under a single API.

> Should we use the GLOG environment variable if that's a standard? Should we try and standardize on one logger?

I believe this is a question that @malfet can answer. Note though that functionality-wise GLOG is a superset of c10's own logger and there are few macros such as `LOG_EVERY_MS()` that work "correctly" only with GLOG (e.g. c10's logger ignores the "every millisecond" part and logs every invocation).

> Looks like JAX relies on the same environment variables as TensorFlow does

Thanks for the pointer! As far as I know virtually every broadly-used library or application has a way for their users to set the verbosity of their output. Some rely on environment variables, some on API calls, but they all ultimately achieve the same thing.

> Given all that, maybe unifying behind one logger is a little too much today, but could we create a PYTORCH_MIN_LOG_LEVEL (or PYTORCH_CPP_MIN_LOG_LEVEL ?) environment variable that controlled logging behavior whether GLOG or the C10 logger was used and promote it as our official solution?

I agree that in short-term migrating to a single logger would require quite a bit of work. Regarding whether we should expose an environment variable (e.g. `PYTORCH_CPP_MIN_LOG_LEVEL`) or an API (e.g. `torch.set_cpp_min_log_level()`), I don't have strong preference. Usability wise an environment variable is easier to deal with; on the other hand (my personal opinion), since we are a library, influencing our runtime behavior should happen through API calls. You can never know in which environment PyTorch is going to be used, maybe the application wants to abstract away its use of PyTorch and does not want its users to influence its program output. I know for instance that over the years OpenMP has suffered a lot from its heavy use of environment variables. But again as long as we all agree on a short-term solution, I am okay with both approaches.

**Comment by mruberry** [comment]
@cbalioglu I completely agree with your analysis and that it's not immediately obvious whether an environment variable or function call is better today (and maybe we'll eventually want both). But if possible I'd suggest we start with the more conceptually named environment variable if that's OK for now. We can always add a function later. I can be convinced otherwise if you think a function is the thing, though.

**Comment by cbalioglu** [comment]
@mruberry totally fine with that suggestion. I can revise this PR and replace the `set_c10_log_level()`/`get_c10_log_level()` API pair with a `PYTORCH_CPP_MIN_LOG_LEVEL` environment variable.

@malfet before proceeding with that change, do you have any feedback? I just want to make sure that we all agree on this solution.

**Comment by cbalioglu** [comment]
@malfet Do you mind sharing your feedback whenever you have time? Most likely you are pretty busy with the v1.11 release, so no rush. Just wanted to make sure that you have seen our latest conversation with @mruberry.

File: c10/util/Logging.cpp
Comment (malfet) [inline c10/util/Logging.cpp:441]: How about something like the follow (or even simple `int FLAGS_caffe2_log_level = getLogLevelFromEnv();` when this flag is defined) 
```suggestion
namespace {
  struct LogLevelInitializer {
    LogLevelInitializer() {
        unsafeSetLogLevelFlagFromEnv();
    }
  } _log_level_initializer;
}
```

**Review by malfet — COMMENTED** [review]
Overall, looks good, but see comments

**Comment by stas00** [comment]
why call it `TORCH_CPP_MIN_LOG_LEVEL` and not `TORCH_CPP_LOG_LEVEL`? 

Pretty much in all logging systems I have seen LOGLEVEL is by definition MIN_LOGLEVEL, since it'll report any level above it.

i.e. you will never want INFO and not WARN/ERROR, no?

File: torch/csrc/Module.cpp
Comment (cbalioglu) [reply torch/csrc/Module.cpp:786]: Although that would be a cleaner solution, there are two issues:

1) We still have to call `SetLogLevelFlagFromEnv` and several other functions when we use GLOG (see GLOG version of `initLogging` [here](https://github.com/pytorch/pytorch/pull/71746/files#diff-a662e72cf0e756a7e539c54f467cafa5f37d8a5e8eb8dd53959dadef54a2cd2cR226).

2) Since the logging unit is shared between PyTorch and caffe2, using a global constructor will interfere when used in caffe2 (which explicitly calls `initCaffe2Logging()`).

File: c10/util/Logging.cpp
Comment (cbalioglu) [reply c10/util/Logging.cpp]: Thanks for the suggestion! Please see the latest revision where I perform a case-insensitive comparison.

File: c10/util/Logging.cpp
Comment (cbalioglu) [reply c10/util/Logging.cpp]: Please see the second point in my comment above.

**Comment by cbalioglu** [comment]
> why call it `TORCH_CPP_MIN_LOG_LEVEL` and not `TORCH_CPP_LOG_LEVEL`?
> 
> Pretty much in all logging systems I have seen LOGLEVEL is by definition MIN_LOGLEVEL, since it'll report any level above it as well. i.e. you will never want INFO and not WARN/ERROR, no?
> 
> edit: I see TF uses MIN, but none of the existing pytorch env vars use it.

I renamed it to `TORCH_CPP_LOG_LEVEL` in the latest revision.

**Review by malfet — APPROVED** [review]

**Comment by rohan-varma** [comment]
Thanks for this change @cbalioglu! 

In addition to allow users to set c10 logging level, this also greatly enhances `TORCH_DISTRIBUTED_DEBUG` feature for OSS users. Before, there was no way for users to configure the logging level other than builiding from scratch. Although, `TORCH_DISTRIBUTED_DEBUG=DETAIL` relies on INFO logging to log certain debug outputs. Allowing OSS users to configure INFO logging makes TORCH_DISTRIBUTED_DEBUG much more practically usable for OSS users as a result (they can use stable versions as opposed to building from scratch).

**Pull request closed without merging**
